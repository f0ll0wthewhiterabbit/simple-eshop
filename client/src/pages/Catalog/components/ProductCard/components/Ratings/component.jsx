import React from 'react'
import PropTypes from 'prop-types'
import ImmutablePropTypes from 'react-immutable-proptypes'
import DeleteIcon from '@material-ui/icons/Delete'

import {
  Stars,
  RatingsRoot,
  RatingWrapper,
  RatingsCount,
  DeleteButton,
  RatingTitle,
  LoadingLabel,
  ErrorLabel,
} from './styles'

const Ratings = ({
  productData,
  changeProductRatingRequest,
  deleteProductRatingRequest,
  ratingsLoadingList,
  ratingsErrorList,
}) => {
  const { _id: id, ratingInfo } = productData
  const { average, votesAmount, currentUserRating } = ratingInfo
  const averageRating = Math.round(average)
  const isUserRatedProduct = Boolean(currentUserRating)
  const isRatingLoading = ratingsLoadingList.indexOf(id) !== -1
  const isErrorInLoad = ratingsErrorList.indexOf(id) !== -1

  const handleRatingChange = (event, userRating) => {
    changeProductRatingRequest({
      productId: id,
      userRating,
    })
  }

  const handleDeleteButtonClick = () => {
    deleteProductRatingRequest(id)
  }

  let userRatingField

  if (isErrorInLoad) {
    userRatingField = (
      <ErrorLabel variant="caption" color="secondary" data-test="ratingError">
        Something went wrong!
      </ErrorLabel>
    )
  } else if (isRatingLoading) {
    userRatingField = (
      <LoadingLabel variant="caption" data-test="ratingLoading">
        Loading...
      </LoadingLabel>
    )
  } else {
    userRatingField = (
      <>
        <Stars
          value={isUserRatedProduct ? currentUserRating : 0}
          name={`simple-controlled-user-${id}`}
          size="small"
          onChange={handleRatingChange}
          readOnly={isUserRatedProduct}
          data-test="userRating"
        />
        {isUserRatedProduct && (
          <DeleteButton
            aria-label="delete"
            size="small"
            onClick={handleDeleteButtonClick}
            data-test="deleteButton"
          >
            <DeleteIcon fontSize="inherit" />
          </DeleteButton>
        )}
      </>
    )
  }

  return (
    <RatingsRoot>
      <RatingWrapper>
        <RatingTitle>Your rating:</RatingTitle>
        {userRatingField}
      </RatingWrapper>
      <RatingWrapper>
        <RatingTitle>Average rating:</RatingTitle>
        <Stars
          value={averageRating}
          name={`simple-controlled-average-${id}`}
          onChange={handleRatingChange}
          readOnly
          size="small"
          data-test="averageRating"
        />
        <RatingsCount active data-test="ratingsCount">
          ({votesAmount})
        </RatingsCount>
      </RatingWrapper>
    </RatingsRoot>
  )
}

Ratings.propTypes = {
  productData: ImmutablePropTypes.recordOf({
    _id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string,
    imageName: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    tags: ImmutablePropTypes.listOf(PropTypes.string),
    ratingInfo: ImmutablePropTypes.recordOf({
      average: PropTypes.oneOfType([
        PropTypes.number.isRequired,
        PropTypes.oneOf([null]).isRequired,
      ]),
      votesAmount: PropTypes.number.isRequired,
      currentUserRating: PropTypes.number,
    }),
  }).isRequired,
  changeProductRatingRequest: PropTypes.func.isRequired,
  deleteProductRatingRequest: PropTypes.func.isRequired,
  ratingsLoadingList: ImmutablePropTypes.listOf(PropTypes.string).isRequired,
  ratingsErrorList: ImmutablePropTypes.listOf(PropTypes.string).isRequired,
}

export default Ratings
