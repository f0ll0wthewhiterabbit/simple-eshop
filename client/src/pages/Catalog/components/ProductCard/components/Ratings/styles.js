import styled from 'styled-components'
import { CardActions, IconButton, Typography } from '@material-ui/core'
import Rating from '@material-ui/lab/Rating'

export const RatingsRoot = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: ${props => props.theme.spacing(2, 2)};
`

export const RatingWrapper = styled(CardActions)`
  display: flex;
  justify-content: flex-start;
  flex-wrap: wrap;
  padding: 0;
  width: 50%;
  min-height: 42px;
  position: relative;
`

export const Stars = styled(Rating)`
  min-height: 18px;

  .MuiRating-iconEmpty {
    color: ${props => props.theme.colors.font.light};
  }

  && {
    margin-left: 0;
  }
`

export const RatingsCount = styled.span`
  margin-left: 6px;
  font-family: ${props => props.theme.font.family};
  color: ${props => props.theme.colors.font.light};
  font-weight: 500;
  font-size: 12px;
`

export const DeleteButton = styled(IconButton)`
  color: ${props => props.theme.colors.font.light};
  position: absolute;
  top: 19px;
  left: 87px;

  && {
    margin-left: 3px;
  }

  &:hover {
    color: ${props => props.theme.colors.font.regular};
  }
`

export const RatingTitle = styled.span`
  width: 100%;
  margin-left: 2px;
  font-family: ${props => props.theme.font.family};
  color: ${props => props.theme.colors.font.light};
  font-weight: 500;
  font-size: 12px;
`

export const LoadingLabel = styled(Typography)`
  font-family: ${props => props.theme.font.family};
  color: ${props => props.theme.colors.font.light};
  font-weight: 500;
  font-size: 12px;
  margin-left: 2px;
`

export const ErrorLabel = styled(Typography)`
  margin-left: 2px;
  font-family: ${props => props.theme.font.family};
  font-weight: 500;
  font-size: 12px;
`
