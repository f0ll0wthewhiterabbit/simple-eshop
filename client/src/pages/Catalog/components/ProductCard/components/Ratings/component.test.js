import React from 'react'
import { shallow } from 'enzyme'
import { Record, List } from 'immutable'

import Ratings from './component'
import findByTestAttr from '../../../../../../utils/findByTestAttr'

jest.mock('../../../../../../utils/formatPrice')

describe('Ratings component', () => {
  let wrapper
  const productData = Record({
    _id: '1',
    title: 'TestTitle',
    description: 'TestDescription',
    imageName: 'TestImageName',
    price: 99,
    tags: List(['firstTag', 'secondTag']),
    ratingInfo: Record({
      average: 5,
      votesAmount: 2,
      currentUserRating: 5,
    })(),
  })()
  const initialProps = {
    productData,
    changeProductRatingRequest: jest.fn(),
    deleteProductRatingRequest: jest.fn(),
    ratingsLoadingList: List(),
    ratingsErrorList: List(),
  }

  const dataTestAverageRating = 'averageRating'
  const dataTestRatingsCount = 'ratingsCount'
  const dataTestUserRating = 'userRating'
  const dataTestRatingError = 'ratingError'
  const dataTestRatingLoading = 'ratingLoading'
  const dataTestDeleteButton = 'deleteButton'

  const generateWrapper = passedProps => {
    const defaultProps = { ...initialProps }
    const props = { ...defaultProps, ...passedProps }

    return shallow(<Ratings {...props} />)
  }

  beforeEach(() => {
    wrapper = generateWrapper()
  })

  afterEach(() => {
    initialProps.changeProductRatingRequest.mockReset()
    initialProps.deleteProductRatingRequest.mockReset()
  })

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should contain correct average rating', () => {
    const averageRating = findByTestAttr(wrapper, dataTestAverageRating)

    expect(averageRating.props().value).toBe(productData.ratingInfo.average)
  })

  it('should contain correct votes amount', () => {
    const ratingsCount = findByTestAttr(wrapper, dataTestRatingsCount)
    const expectedAmount = `(${productData.ratingInfo.votesAmount})`

    expect(ratingsCount.text()).toBe(expectedAmount)
  })

  it('should contain correct user rating', () => {
    const userRating = findByTestAttr(wrapper, dataTestUserRating)

    expect(userRating.props().value).toBe(productData.ratingInfo.currentUserRating)
  })

  it('should contain delete user rating button', () => {
    const deleteButton = findByTestAttr(wrapper, dataTestDeleteButton)

    expect(deleteButton).toHaveLength(1)
  })

  it('should handle delete user rating', () => {
    const deleteButton = findByTestAttr(wrapper, dataTestDeleteButton)
    deleteButton.simulate('click')

    expect(initialProps.deleteProductRatingRequest).toHaveBeenCalledTimes(1)
    expect(initialProps.deleteProductRatingRequest).toHaveBeenCalledWith(productData._id)
  })

  it('should handle change user rating', () => {
    const newUserRating = 1
    const userRating = findByTestAttr(wrapper, dataTestUserRating)
    userRating.simulate('change', {}, newUserRating)

    expect(initialProps.changeProductRatingRequest).toHaveBeenCalledTimes(1)
    expect(initialProps.changeProductRatingRequest).toHaveBeenCalledWith({
      productId: productData._id,
      userRating: newUserRating,
    })
  })

  it('should contain correct rating error message', () => {
    wrapper = generateWrapper({ ratingsErrorList: List([productData._id]) })
    const ratingError = findByTestAttr(wrapper, dataTestRatingError)

    expect(ratingError).toHaveLength(1)
    expect(ratingError.text()).toBe('Something went wrong!')
  })

  it('should contain correct rating loading message', () => {
    wrapper = generateWrapper({ ratingsLoadingList: List([productData._id]) })
    const ratingLoading = findByTestAttr(wrapper, dataTestRatingLoading)

    expect(ratingLoading).toHaveLength(1)
    expect(ratingLoading.text()).toBe('Loading...')
  })

  describe('without user rating', () => {
    beforeEach(() => {
      const productDataWithoutUserRating = Record({
        _id: '1',
        title: 'TestTitle',
        description: 'TestDescription',
        imageName: 'TestImageName',
        price: 99,
        tags: List(['firstTag', 'secondTag']),
        ratingInfo: Record({
          average: 5,
          votesAmount: 1,
        })(),
      })()
      wrapper = generateWrapper({ productData: productDataWithoutUserRating })
    })

    it(`shouldn't contain user rating`, () => {
      const userRating = findByTestAttr(wrapper, dataTestUserRating)

      expect(userRating.props().value).toBe(0)
    })

    it(`shouldn't contain delete user rating button`, () => {
      const deleteButton = findByTestAttr(wrapper, dataTestDeleteButton)

      expect(deleteButton).toHaveLength(0)
    })
  })
})
