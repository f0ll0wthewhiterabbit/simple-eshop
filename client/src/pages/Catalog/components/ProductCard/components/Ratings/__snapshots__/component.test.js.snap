// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ratings component should render correctly 1`] = `
<styled.div>
  <Styled(WithStyles(ForwardRef(CardActions)))>
    <styled.span>
      Your rating:
    </styled.span>
    <Styled(WithStyles(ForwardRef(Rating)))
      data-test="userRating"
      name="simple-controlled-user-1"
      onChange={[Function]}
      readOnly={true}
      size="small"
      value={5}
    />
    <Styled(WithStyles(ForwardRef(IconButton)))
      aria-label="delete"
      data-test="deleteButton"
      onClick={[Function]}
      size="small"
    >
      <DeleteIcon
        fontSize="inherit"
      />
    </Styled(WithStyles(ForwardRef(IconButton)))>
  </Styled(WithStyles(ForwardRef(CardActions)))>
  <Styled(WithStyles(ForwardRef(CardActions)))>
    <styled.span>
      Average rating:
    </styled.span>
    <Styled(WithStyles(ForwardRef(Rating)))
      data-test="averageRating"
      name="simple-controlled-average-1"
      onChange={[Function]}
      readOnly={true}
      size="small"
      value={5}
    />
    <styled.span
      active={true}
      data-test="ratingsCount"
    >
      (
      2
      )
    </styled.span>
  </Styled(WithStyles(ForwardRef(CardActions)))>
</styled.div>
`;
